Add a MongoDB client extension (#3022)
quarkusio/quarkus: docs/src/main/asciidoc/mongo-guide.adoc

In this example, we will create an application to manage a list of fruits.

First, let's create the `Fruit` bean as follows:

[source,java]
----
package org.acme.rest.json;

import java.util.Objects;

public class Fruit {

    private String name;
    private String description;

    public Fruit() {
    }

    public Fruit(String name, String description) {
        this.name = name;
        this.description = description;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Fruit)) {
            return false;
        }

        Fruit other = (Fruit) obj;

        return Objects.equals(other.name, this.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.name);
    }
}
----

Nothing fancy. One important thing to note is that having a default constructor is required by the JSON serialization layer.

> Now create a `org.acme.rest.json.FruitService` that will be the business layer of our application and store/load the fruits from the mongo database.

Author: Here we cannot say MongoDB database, as DB is for database.

                list.add(fruit);
            }
        } finally {
            cursor.close();
        }
        return list;
    }

    public void add(Fruit fruit){
        Document document = new Document()
                .append("name", fruit.getName())
                .append("description", fruit.getDescription());
        getCollection().insertOne(document);
    }

    private MongoCollection getCollection(){
        return mongoClient.getDatabase("fruit").getCollection("fruit");
    }
}
----

Now, edit the `org.acme.rest.json.FruitResource` class as follows:

[source,java]
----
@Path("/fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class FruitResource {

    @Inject FruitService fruitService;

    @GET
    public List<Fruit> list() {
        return fruitService.list();
    }

    @POST
    public List<Fruit> add(Fruit fruit) {
        fruitService.add(fruit);
        return list();
    }
}
----

The implementation is pretty straightforward and you just need to define your endpoints using the JAX-RS annotations and use the FruitService to list/add new fruits.

> == Configuring the mongo database

Author: Same comment as above, when followed with database, it should be mongo.

> The main property to configure is the URL to access to MongoDB, almost all Mongo configuration can be put on the connection URI so we advise you to do so, you can find more information on the Mongo documentation site : https://docs.mongodb.com/manual/reference/connection-string/

Author: Sentence rewritten.

A sample configuration should look like this:
[source,properties]
----
# configure mongo client for a replica set of two nodes
quarkus.mongodb.connectionString = mongodb://mongo1:27017,mongo2:27017
----

If you need more configuration properties, there is a full list at the end of this guide.

> == Running a Mongo Database

Reviewer: ```suggestion
== Running a MongoDB Database
```

> As by default, `MongoClient` is configure to access a local Mongo database on port 27017 (the default Mongo port), if you have a local running database on this port, there is nothing more to do before being able to test it!

Reviewer: ```suggestion
As by default, `MongoClient` is configure to access a local MongoDB database on port 27017 (the default MongoDB port), if you have a local running database on this port, there is nothing more to do before being able to test it!
```

> If you want to use Docker to run a Mongo database, you can use the following command to launch one:

Reviewer: ```suggestion
If you want to use Docker to run a MongoDB database, you can use the following command to launch one:
```

[source,shell]
----
docker run -ti --rm -p 27017:27017 mongo:4.0
----

== Creating a frontend

Now let's add a simple web page to interact with our `FruitResource`.
Quarkus automatically serves static resources located under the `META-INF/resources` directory.

> In the `src/main/resources/META-INF/resources` directory, add a `fruits.html` file with the content from this {quickstarts-blob-url}/rest-json/src/main/resources/META-INF/resources/fruits.html[fruits.html] file in it.

Reviewer: I don't think you're referring to the right quickstart?

Author: it's copying the same file, but yes, would be better to provide the file in the right quickstart. Keeping this comment open until we have the right url.

You can now interact with your REST service:

 * start Quarkus with `./mvnw compile quarkus:dev`
 * open a browser to `http://localhost:8080/fruits.html`
 * add new fruits to the list via the form

> == Reactive Mongo Client

Reviewer: ```suggestion
== Reactive MongoDB Client
```

> A reactive Mongo Client is included in Quarkus.

Reviewer: ```suggestion
A reactive MongoDB Client is included in Quarkus.
```

> Using it is as easy as using the classic Mongo Client. You can rewrite the previous example to use it like the following.

Reviewer: ```suggestion
Using it is as easy as using the classic MongoDB Client. You can rewrite the previous example to use it like the following.
```

            fruit.setDescription(doc.getString("description"));
            return fruit;
        }).toList().run();
    }

    public CompletionStage<Void> add(Fruit fruit){
        Document document = new Document()
                .append("name", fruit.getName())
                .append("description", fruit.getDescription());
        return getCollection().insertOne(document);
    }

    private ReactiveMongoCollection<Document> getCollection(){
        return mongoClient.getDatabase("fruit").getCollection("fruit");
    }
}
----

[source,java]
----
package org.acme.rest.json;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import java.util.List;
import java.util.concurrent.CompletionStage;

@Path("/reactive_fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class ReactiveFruitResource {

    @Inject ReactiveFruitService fruitService;

    @GET
    public CompletionStage<List<Fruit>> list() {
        return fruitService.list();
    }

    @POST
    public CompletionStage<List<Fruit>>  add(Fruit fruit) {
        fruitService.add(fruit);
        return list();
    }
}
----

> == Simplifying Mongo Client usage

Reviewer: Same comment about MongoDB but I suppose it depends on how you want to name the extension in the end.

Author: Here as it's not followed with database, it should be MongoDB.

One of the drawback of using directly the Mongo Client, is that you need to manually transform your objects to/from a Mongo `Document`.

> To facilitate this there is two ways.

Reviewer: ```suggestion
To facilitate this there are two ways.
```
?

Author: section dropped.

=== Using a Bson Codec

> By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a Mongo `Document` automatically.

Reviewer: ```suggestion
By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a MongoDB `Document` automatically.
```

> First you need to create a Bson `Codec` that will tells Bson how to transform your entity to/from a Mongo `Document`.

Reviewer: ```suggestion
First you need to create a Bson `Codec` that will tell Bson how to transform your entity to/from a MongoDB `Document`.
```

> Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a Mongo identifier), if not we would have used a `Codec` instead.

Reviewer: ```suggestion
Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a MongoDB identifier), if not we would have used a `Codec` instead.
```

> More information on the codec documentation : https://mongodb.github.io/mongo-java-driver/3.10/bson/codecs.

Reviewer: ```suggestion
More information in the codec documentation: https://mongodb.github.io/mongo-java-driver/3.10/bson/codecs.
```

Author: fixed

        doc.put("name", fruit.getName());
        doc.put("description", fruit.getDescription());
        documentCodec.encode(writer, doc, encoderContext);
    }

    @Override
    public Class<Fruit> getEncoderClass() {
        return Fruit.class;
    }

    @Override
    public Fruit generateIdIfAbsentFromDocument(Fruit document) {
        if (!documentHasId(document)) {
            document.setId(UUID.randomUUID().toString());
        }
        return document;
    }

    @Override
    public boolean documentHasId(Fruit document) {
        return document.getId() != null;
    }

    @Override
    public BsonValue getDocumentId(Fruit document) {
        return new BsonString(document.getId());
    }

    @Override
    public Fruit decode(BsonReader reader, DecoderContext decoderContext) {
        Document document = documentCodec.decode(reader, decoderContext);
        Fruit fruit = new Fruit();
        if (document.getString("id") != null) {
            fruit.setId(document.getString("id"));
        }
        fruit.setName(document.getString("name"));
        fruit.setDescription(document.getString("description"));
        return fruit;
    }
}
----

> Then you need to create a `CodecProvider` to linked this `Codec` to the `Fruit` class.

Reviewer: ```suggestion
Then you need to create a `CodecProvider` to link this `Codec` to the `Fruit` class.
```

Author: fixed.

        MongoCursor<Fruit> cursor = getCollection().find().iterator();

        try {
            while (cursor.hasNext()) {
                list.add(cursor.next());
            }
        } finally {
            cursor.close();
        }
        return list;
    }

    public void add(Fruit fruit){
        getCollection().insertOne(fruit);
    }

    private MongoCollection<Fruit> getCollection(){
        return mongoClient.getDatabase("fruit").getCollection("fruit", Fruit.class);
    }
}
----

== Building a native executable

You can use the Mongo client in a native executable.

If you want to use SSL/TLS encryption, you need to add these properties in your `application.properties`:
----
quarkus.mongodb.tls=true
quarkus.mongodb.tls-insecure=true #only if TLS certificat cannot be validated
----

You can then build a native executable with the usual command `./mvnw package -Pnative`.

Running it is as simple as executing `./target/mongo-client-1.0-SNAPSHOT-runner`.

You can then point your browser to `http://localhost:8080/fruits.html` and use your application.

[WARNING]
====
Currently, Quarkus doesn't support the `mongodb+srv` protocol in native mode.
====

== Conclusion

> Accessing a Mongo database from a Mongo Client is easy with Quarkus as it provides configuration and native support for it.

Reviewer: ```suggestion
Accessing a MongoDB database from a Mongo Client is easy with Quarkus as it provides configuration and native support for it.
```

== Connection properties
`quarkus.mongodb.connection-string`:: (e.g. `mongodb://mongo1:27017,mongo2:27017`)
MongoDB connection string, this is a URI where a lot of options can be passed : https://docs.mongodb.com/manual/reference/connection-string/

`quarkus.mongodb.hosts`::

> Coma separated lits of hosts.

Reviewer: ```suggestion
Comma separated list of hosts.
```

Author: fixed.

`quarkus.mongodb.application-name`::
Application name.

`quarkus.mongodb.max-pool-size`::
Maximum number of connections in the connection pool.

`quarkus.mongodb.min-pool-size`::

> Minumum number of connections in the connection pool.

Reviewer: ```suggestion
Minimum number of connections in the connection pool.
```

Author: Also had to be fixed in the javadoc.

`quarkus.mongodb.max-connection-idle-time`::

> Maximum idle time in milliseconds of a pooled connection. A connection that exceeds this limit will be closed.

Reviewer: Should they be `Duration` instead? Same question for all these variables.

(we have a note to include if this is the case)

Author: Wondering about it. This is the MongClient configuration passed almost directly to the configuration. But yes, I would prefer using duration.

Author: Fixed - used duration everywhere.

`quarkus.mongodb.max-connection-life-time`::
Maximum life time in milliseconds of a pooled connection. A connection that exceeds this limit will be closed.

`quarkus.mongodb.wait-queue-timeout`::
The maximum wait time in milliseconds that a thread may wait for a connection to become available.

`quarkus.mongodb.maintenance-frequency`::
Configures the time period in milliseconds between runs of the maintenance job. Default is 0.

`quarkus.mongodb.maintenance-initial-delay`::
Configures period of time in milliseconds to wait before running the first maintenance job on the connection pool.

`quarkus.mongodb.wait-queue-multiple`::

> This multiplier, multiplied with the `maxPoolSi` setting, gives the maximum number of threads that may be waiting for a connection to become available from the pool. All further threads will get an exception right away.

Reviewer: maxPoolSi?

Author: ze :-D

`quarkus.mongodb.connect-timeout`::
How long in milliseconds a connection can take to be opened before timing out.

`quarkus.mongodb.socket-timeout`::
How long in milliseconds a send or receive on a socket can take before timing out.

`quarkus.mongodb.tls-insecure`::

> If connecting with TLS, this option enables insecure TLS connections. Default to false.

Reviewer: ```suggestion
If connecting with TLS, this option enables insecure TLS connections. Defaults to false.
```

`quarkus.mongodb.tls`::

> Whether to connect using TLS. Default to false.

Reviewer: ```suggestion
Whether to connect using TLS. Defaults to false.
```

Configures the journal writing aspect.
If set to `true` (the default): the driver waits for the server to group commit to the journal file on disk.
If set to `false`: the driver does not wait for the server to group commit to the journal file on disk.

`quarkus.mongodb.write-concern.w`::
When set, the driver adds `w: wValue` to all write commands. It requires `safe` to be `true`.
The value is typically a number, but can also be the `majority` string.

`quarkus.mongodb.write-concern.retry-writes`::
If set to `true`, the driver will retry supported write operations if they fail due to a network error. The default is `false`.

`quarkus.mongodb.write-concern.w-timeout`::
When set, the driver adds `wtimeout : ms`  to all write commands. It requires `safe` to be `true`. Time given in ms.

=== Credentials
`quarkus.mongodb.credentials.username`::
Configures the username.

`quarkus.mongodb.credentials.password`::
Configures the password.

`quarkus.mongodb.credentials.auth-mechanism`::
Configures the authentication mechanism to use if a credential was supplied.
The default is unspecified, in which case the client will pick the most secure mechanism available based on the

> sever version. For the GSSAPI and MONGODB-X509 mechanisms, no password is accepted, only the username.

Reviewer: ```suggestion
server version. For the GSSAPI and MONGODB-X509 mechanisms, no password is accepted, only the username.
```

Supported values: `MONGO-CR|GSSAPI|PLAIN|MONGODB-X509`

`quarkus.mongodb.credentials.auth-source`::
Configures the source of the authentication credentials.

> This is typically the database that the credentials have been created. The value defaults to the database

Reviewer: `the database that the credentials have been created` hmmm, `that` is weird here.

Author: Updated. This was actually in the upstream code...
